options.txt
# This is a sample options file for Uroboros.
# For the avoidance of doubt, this file is released into the public domain,
# just like Uroboros itself.
# The format for options is `option = value`.  Whitespace is not optional.
# Boolean options can be enabled with "y" or "yes" (no quotes) or disabled
# with "n" or "no".
# This file is read line-by-line, and as such only one option is permitted per
# line.  Lines that begin with a # will be ignored, as will blank lines.
# All lines, including comment lines, must be 80 characters long or less.
# If the same option is set more than once, later values will take precedence.


# Key bindings


# Key bindings can be defined using the character that gets output from
# pressing that key.  A capital letter will require you to press the SHIFT key
# in-game.  As a special case, Uroboros will accept KEY_LEFT, KEY_RIGHT,
# KEY_UP, and KEY_DOWN for the arrow keys and SPACE for the space bar.
move_up = KEY_UP
move_down = KEY_DOWN
move_left = KEY_LEFT
move_right = KEY_RIGHT

# Remove the # at the beginning of the following lines to enable WASD movement:
#move_up = w
#move_down = s
#move_left = a
#move_right = d

# Old-schoolers may prefer these vi-like key bindings:
#move_left = h
#move_down = j
#move_up = k
#move_right = l


# These keys will cause your snake to speed up or slow down.
speed_up = =
slow_down = -


# Miscellaneous keybindings.
pause_key = p
quit_key = q
show_version = v
show_license = c


# Display settings


# Disabling color will cause Uroboros to use an alternative character set which
# should hopefully be more visible on terminals which don't support color.
color = yes


CHANGELOG.md
# Changelog

This file records the concise(-ish) history of the Uroboros project.

The format of this file is based on the
[Keep a Changelog 1.0.0](https://keepachangelog.com/en/1.0.0/) convention.

An effort has been made for this project to follow the
[Semantic Versioning 2.0.0](https://semver.org/spec/v2.0.0) version numbering
standard.  Prerelease versions of Uroboros are referred to as "dev builds" and
indicated by a "-dev" suffix followed by a number.  Dev builds are generally
not documented in this file unless special circumstances require it.


## [0.3.0]

### Notes

- Windows support has been dropped due to unresolvable segmentation faults
  whose origins remain a mystery.

### Fixed

- Lines read in from the options file are now reset after each line read, to
  prevent assigning options twice under certain circumstances.

- The "hilite\_apples" option has been removed in favor of simply hiding the
  terminal cursor on startup.

### Added

- Debug builds now contain some `assert` calls used to verify that the memory
  for the player character's head and tail has been allocated properly.

### Code Changes

- Strings used when reading in the options file are now statically allocated
  rather than being dynamically allocated at runtime.  They should also use
  less memory this way.

- The "key" string, used when reading in the names of options to be set, has
  been shrunk down to 12 characters, the length of the longest option name.


## [0.3.0-dev.2] - 2022-12-03

### Notes

- In order to protect users who are subject to tyrannical governments, the
  license agreement has been changed from the previous informal license to the
  Unlicense.

### Added

- Certain program settings, such as key bindings and display settings, can now
  be set by the user in an options file, included with Uroboros.  The game no
  longer needs to be recompiled in order to be reconfigured.

- If the player's terminal is shorter than the height of the license agreement
  disclosure window, the window can now be scrolled up and down.

### Fixed

- The message mocking people who die with low scores now appropriately displays
  whenever the player's score is impossibly low, not just when it's under 4
  points.

### Code Changes

- A new variable type `flag_t` will now be used for bitmask flags.

- Player status is now stored in a `flag_t` named `status` (originally named
  `dead`) and now uses the named constants `DEAD` and `CHEATED` to determine
  the respective player statuses.


## [0.2.2] - 2022-10-18

### Added

- The game now displays a message on the high scores table if it detects
  cheating.

- Debug builds now have a small message informing you of the identity of a key
  that has been pressed.

### Changed

- The version number is now always displayed when the game is started, rather
  than only being displayed on debug or prerelease versions.

- The player's score is now displayed with the high scores list, rather than
  separately from it.

### Fixed

- On Windows, the player's score data is output a second time after entering
  their name, to fix a problem introduced by the Windows terminal clearing all
  text data after the name input.

- The high scores list is now output even if the player cheated.

- User input is now flushed before trying to enter a player name.


## [0.2.1] - 2022-10-13

### Added

- An in-game high scores window will now be displayed, rather than outputting
  the high scores to the terminal after the game closes.

### Changed

- Player names are no longer stored with a bunch of dots in the high scores
  file.

- The helpful text is now cleared off of the screen when the game begins.

### Fixed

- The terminal will no longer refuse to echo input after quitting the game.
  I still don't know why that happened, but it's fixed now.

- The first apple will no longer spawn on the same line as the version number,
  to prevent the "Eat these" message from interfering with the version number
  and vice-versa.

- The timer now doesn't start until the player presses a key, so they can wait
  for as long as they want to start playing without their high score being
  affected.


## [0.2.0] - 2022-10-13

### Added

- High scores!  Qualifying players can now add their name to a "top 10" list
  that they can use to gloat to other players!

- Debug output will now include the player's ratio along with their length.

### Changed

- The license disclosure window is now centered.


## [0.1.0] - 2022-09-19

### Added

- The game can now be compiled without color if desired.

- The player can now increase or decrease the game's speed on-the-fly with the
  \- and + keys.

- On debug builds, the player can now increase or decrease the size of the
  snake by holding shift while pressing the - or + keys.

### Changed

- The cursor now always moves over apples.  The `HILITE_APPLE` flag now only
  determines if the apple's colors are reversed to help hide the cursor.

- The snake is now green.

- Version numbers are now only displayed on debug builds or prerelease
  versions.

### Removed

- Several unused functions previously intended to be used to handle the linked
  lists used to represent the snake have been removed.

### Fixed

- Uses of the `true` keyword, which is not standard across C compilers, have
  been removed.

- "curses.h" has been used instead of "ncurses.h" to allow for the use of
  standards other than ncurses.

- Game speed is now appropriately determined according to the `CLOCKS_PER_SEC`
  constant, allowing for cross-compatibility between systems.  This has been
  used to allow a fairly consistent default speed of 10 frames per second.


src/options.c
// This is free and unencumbered software released into the public domain.
//
// For more information, please refer to the Unlicense, which should be
// included with the original source distribution or at <https://unlicense.org>
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
// AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
// ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


// This file defines functions declared in options.h


#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "curses.h"

#include "globals.h"
#include "options.h"


// Initialize all options to their default values.
void init_options()
{

    move_up      = KEY_UP;
    move_down    = KEY_DOWN;
    move_left    = KEY_LEFT;
    move_right   = KEY_RIGHT;
    speed_up     = '=';
    slow_down    = '-';
#ifdef DEBUG
    debug_grow   = '+';
    debug_shrink = '_';
#endif
    quit_key     = 'q';
    pause_key    = 'p';
    show_version = 'v';
    show_license = 'c';

    color = 1;

    hiscore_path = "hiscore";

}


// Attempts to assign a key binding to `target` based on the input `value`.
// Returns a 1 on success or a 0 on failure.
int assign_keybind( int *target, const char *value )
{

    int succ = 0;

    // Generally, we only care about the first character of `value`.  However,
    // as a special case, Uroboros allows arrow key inputs or "SPACE" to
    // indicate the space bar, so it behooves us to check.
    
    // If the input `value` is only one character long, that's a freebie for
    // us.
    if( strlen( value ) == 1 )
    {
        *target = value[0];
        succ = 1;

    }
    // If `value` starts with 'S', it might be the "SPACE" keyword used by
    // Uroboros to indicate the user wants to assign the keybind to the
    // spacebar.
    else if( value[0] == 'S' )
    {

        if( !strcmp( value, "SPACE" ) )
        {

            *target = ' ';
            succ = 1;

        }

    }
    // If `value` starts with 'K', that means that it may be one of the `KEY_*`
    // constants defined by curses, which in our case means it may be one of
    // the arrow keys which Uroboros accepts.
    else if( value[0] == 'K' )
    {

        // Check each of the names in order from shortest to longest.
        if( !strcmp( value, "KEY_UP" ) )
        {

            *target = KEY_UP;
            succ = 1;

        }
        else if( !strcmp( value, "KEY_DOWN" ) )
        {

            *target = KEY_DOWN;
            succ = 1;

        }
        else if( !strcmp( value, "KEY_LEFT" ) )
        {

            *target = KEY_LEFT;
            succ = 1;

        }
        else if( !strcmp( value, "KEY_RIGHT" ) )
        {

            *target = KEY_RIGHT;
            succ = 1;

        }

    } // else if( value[0] == 'K' )

    return succ;

} // int assign_keybind( int*, const char* )


// Reads in options from a file at the given path.  This function will print
// warning messages to stdio if it reads an invalid line from the file.
// Returns a positive error count if any such warnings were printed, 0
// otherwise.  Returns -1 if the options file was not found.
int read_options_file( const char *path )
{

    int err = 0;

    FILE *options = fopen( path, "r" );

    // Abort if the file is not found and return -1.
    if( options == NULL )
    {

        return -1;

    }

    // Read the file.
    char line[81];
    char key[13];
    char value[61];
    line[80]  = '\0';
    key[12]   = '\0';
    value[60] = '\0';
    size_t count = 0;
    int pair;
    size_t length;
    int keybind;
    char check_keybind;
    while( !feof( options ) )
    {

        line[0] = '\0';
        check_keybind = 0;
        count = count + 1;

        // Read up to 80 characters of the given line.
        fgets( line, 80, options );

        // If the line is blank or starts with a '#', skip it.
        length = strlen( line );
        if( length <= 1 || line[0] == '#' )
        {

            continue;

        }

        // Non-comment lines follow a predictable format, so we can use
        // `sscanf` to get key/value pairs for each option.
        pair = sscanf( line, "%s = %s", key, value );

        // `pair` should evaluate to 2 if we successfully got a key/value
        // pair.  If not, print a warning message and skip this line.
        if( pair != 2 )
        {

            printf( "WARNING: Could not read line %lu in %s:\n%s\n",
                    count, path, line );

            err = err + 1;

            continue;

        }

        // Now we check each of the possible name values that could have been
        // read into `key` and assign a value appropriately.
        
        // Key bindings 


        if( !strcmp( key, "move_up" ) )
        {

            keybind = assign_keybind( &move_up, value );
            check_keybind = 1;

        }
        else if( !strcmp( key, "move_down" ) )
        {

            keybind = assign_keybind( &move_down, value );
            check_keybind = 1;

        }
        else if( !strcmp( key, "move_left" ) )
        {

            keybind = assign_keybind( &move_left, value );
            check_keybind = 1;

        }
        else if( !strcmp( key, "move_right" ) )
        {

            keybind = assign_keybind( &move_right, value );
            check_keybind = 1;

        }
        else if( !strcmp( key, "speed_up" ) )
        {
            
            keybind = assign_keybind( &speed_up, value );
            check_keybind = 1;

        }
        else if( !strcmp( key, "slow_down" ) )
        {

            keybind = assign_keybind( &slow_down, value );
            check_keybind = 1;

        }
        else if( !strcmp( key, "pause_key" ) )
        {

            keybind = assign_keybind( &pause_key, value );
            check_keybind = 1;

        }
        else if( !strcmp( key, "quit_key" ) )
        {

            keybind = assign_keybind( &quit_key, value );
            check_keybind = 1;

        }
        else if( !strcmp( key, "show_version" ) )
        {

            keybind = assign_keybind( &show_version, value );
            check_keybind = 1;

        }
        else if( !strcmp( key, "show_license" )
              || !strcmp( key, "show_licence" ) )
        {

            keybind = assign_keybind( &show_license, value );
            check_keybind = 1;

        }
        // Debug key bindings don't trigger a warning on non-debug builds, but
        // also don't assign any values.
        else if( !strcmp( key, "debug_grow" ) )
        {

#ifndef DEBUG
            continue;
#else
            keybind = assign_keybind( &debug_grow, value );
            check_keybind = 1;
#endif

        }
        else if( !strcmp( key, "debug_shrink" ) )
        {

#ifndef DEBUG
            continue;
#else
            keybind = assign_keybind( &debug_shrink, value );
            check_keybind = 1;
#endif

        }

        // If a keybind option was read, check if it was successful.
        if( check_keybind )
        {

            // If the keybind was unsuccessful, print a warning.
            if( !keybind )
            {

                printf( "WARNING: Invalid keybind at line %lu in %s:\n%s\n",
                        count, path, line );
                err = err + 1;

            }

            // Either way, we are finished and can move on to the next line.
            continue;

        }
        else if( !strcmp( key, "color" ) || !strcmp( key, "colour" ) )
        {

            if( value[0] == 'y' || value[0] == 'Y' )
            {

                color = 1;

            }
            else if( value[0] == 'n' || value[0] == 'N' )
            {

                color = 0;

            }
            else
            {

                printf( "WARNING: Invalid boolean option at line %lu in %s:",
                        count, path );
                printf( "\n%s\n", line );
                err = err + 1;

            }

            continue;

        }

        // If none of the above option checks have passed, the line is invalid
        // and we need to nag the user.
        else
        {

            printf( "WARNNIG: Invalid or unknown key/value pair at line %lu",
                    count );
            printf( " in %s:\n%s\n", path, line );
            err = err + 1;

        }

    } // while( !eof( options ) )

    fclose( options );

#ifdef DEBUG
    puts( "DEBUG: Reading options file resulted in assigned options:" );

# define show_keybind( name, var ) printf( "%s:\t%d/%c\n", name, var, var )
# define show_boolean( name, val, mask ) printf( "%s:\t%c\n", name, \
        mask & val ? 'Y' : 'N' )

    show_keybind( "move_up", move_up );
    show_keybind( "move_down", move_down );
    show_keybind( "move_left", move_left );
    show_keybind( "move_right", move_right );
    show_keybind( "speed_up", speed_up );
    show_keybind( "slow_down", slow_down );
    show_keybind( "debug_grow", debug_grow );
    show_keybind( "debug_shrink", debug_shrink );
    show_keybind( "pause_key", pause_key );
    show_keybind( "quit_key", quit_key );
    show_keybind( "show_version", show_version );
    show_keybind( "show_license", show_license );
    show_boolean( "color", 1, color );

# undef show_keybind
# undef show_boolean

#endif

    return err;

} // int read_options_file( const char* )
